consensus/src/chained_bft/event_processor.rs

```diff
@@ -355,53 +367,116 @@ impl<T: Payload, P: ProposerInfo> EventProcessor<T, P> {
+    async fn sync_up(
+        &mut self,
+        sync_info: &SyncInfo,
+        preferred_peer: Option<Author>,
+    ) -> failure::Result<()> {
+        let current_hqc_round = self
             .block_store
             .highest_quorum_cert()
             .certified_block_round();
-        let new_round_highest_quorum_cert_round = timeout_msg
-            .sync_info()
-            .highest_quorum_cert()
-            .certified_block_round();
+        let sync_info_hqc_round = sync_info.highest_quorum_cert().certified_block_round();
 
-        if current_highest_quorum_cert_round < new_round_highest_quorum_cert_round {
-            // The timeout message carries a QC higher than what this node has seen before:
-            // run state synchronization.
-            let deadline = self.pacemaker.current_round_deadline();
-            match self
-                .sync_manager
-                .sync_to(
-                    deadline,
-                    SyncMgrContext::new(
-                        timeout_msg.sync_info(),
-                        timeout_msg.author(),
-                    ),
-                )
-                .await
-                {
-                    Ok(()) => debug!(
-                        "Successfully added new highest quorum certificate at round {} from old round {}",
-                        new_round_highest_quorum_cert_round, current_highest_quorum_cert_round
-                    ),
-                    Err(e) => warn!(
-                        "Unable to insert new highest quorum certificate {} from old round {} due to {:?}",
-                        timeout_msg.sync_info().highest_quorum_cert(),
-                        current_highest_quorum_cert_round,
-                        e
-                    ),
-                }
+        if current_hqc_round >= sync_info_hqc_round {
+            return Ok(());
         }
-        self.pacemaker
-            .process_remote_timeout(timeout_msg.pacemaker_timeout().clone())
-            .await;
+
+        debug!(
+            "Starting sync: current_hqc_round = {}, sync_info_hqc_round = {}",
+            current_hqc_round, sync_info_hqc_round,
+        );
+        let deadline = self.pacemaker.current_round_deadline();
+        let sync_mgr_context = SyncMgrContext::new(sync_info, preferred_peer);
+        self.sync_manager
+            .sync_to(deadline, sync_mgr_context)
+            .await
+            .map_err(|e| {
+                warn!(
+                    "Fail to sync up to HQC @ round {}: {:?}",
+                    sync_info_hqc_round, e
+                );
+                e
+            })?;
+        debug!("Caught up to HQC at round {}", sync_info_hqc_round);
+        Ok(())
     }
 
     pub async fn process_sync_info_msg(&mut self, sync_info: SyncInfo) {
         debug!("Received a sync info msg: {}", sync_info);
+        // First bring missing dependencies, then update pacemaker.
+        if self.sync_up(&sync_info, None).await.is_err() {
```

Comment:
Is this the only case we use None as preferred_peer? I think use the author of the message might be better?

Reply:
You know what, I looked at our code again and we can indeed unify the fetch_qc / sync_to parts in event processor s.t. everything would go through the `sync_up` function. Will send an update!

Reply:
Hmm sure why not, let's try the author of the message as the preferred peer.